# 프로젝트 (1) load_digits : 손글씨를 분류

In [1]:
import sklearn
print(sklearn.__version__)

1.0


In [2]:
# (1) 필요한 모듈 import하기
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [5]:
# (2) 데이터 준비
digits = load_digits()
print(dir(digits))

['DESCR', 'data', 'feature_names', 'frame', 'images', 'target', 'target_names']


In [6]:
# (3) 데이터 이해하기
digits_data = digits.data    # Feature Data 지정
print(digits_data.shape)

(1797, 64)


In [7]:
digits_data[0]

array([ 0.,  0.,  5., 13.,  9.,  1.,  0.,  0.,  0.,  0., 13., 15., 10.,
       15.,  5.,  0.,  0.,  3., 15.,  2.,  0., 11.,  8.,  0.,  0.,  4.,
       12.,  0.,  0.,  8.,  8.,  0.,  0.,  5.,  8.,  0.,  0.,  9.,  8.,
        0.,  0.,  4., 11.,  0.,  1., 12.,  7.,  0.,  0.,  2., 14.,  5.,
       10., 12.,  0.,  0.,  0.,  0.,  6., 13., 10.,  0.,  0.,  0.])

In [9]:
digits_label = digits.target   # Target Data 지정
print(digits_label.shape)
digits_label                    

(1797,)


array([0, 1, 2, ..., 8, 9, 8])

In [10]:
digits.target_names      # Target Names 출력

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [11]:
print(digits.DESCR)     # 데이터 Describe

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

In [13]:
# (4) train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(digits_data,
                                                   digits_label,
                                                   test_size=0.2,
                                                   random_state=10)
print('X_train 개수: ', len(X_train),', X_test 개수: ', len(X_test))

X_train 개수:  1437 , X_test 개수:  360


In [14]:
# (5) 다양한 모델로 학습시켜보기

# Decision Tree 사용해 보기
from sklearn.tree import DecisionTreeClassifier
decision_tree = DecisionTreeClassifier(random_state = 40)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)
y_pred

array([5, 8, 5, 4, 1, 2, 4, 3, 3, 0, 8, 7, 0, 1, 8, 6, 9, 7, 9, 7, 1, 3,
       6, 7, 8, 8, 5, 3, 5, 9, 3, 3, 7, 3, 4, 1, 9, 2, 5, 4, 2, 1, 0, 9,
       2, 3, 6, 1, 9, 8, 4, 3, 8, 4, 9, 5, 9, 7, 8, 0, 4, 3, 1, 4, 7, 9,
       0, 7, 1, 3, 4, 3, 3, 8, 0, 7, 3, 6, 5, 2, 0, 8, 8, 0, 1, 1, 2, 8,
       8, 8, 3, 6, 3, 4, 7, 9, 2, 2, 9, 2, 5, 9, 8, 0, 6, 2, 8, 0, 6, 7,
       7, 3, 9, 1, 5, 4, 6, 0, 8, 8, 1, 2, 7, 2, 4, 5, 2, 3, 8, 7, 7, 6,
       2, 5, 4, 9, 3, 7, 6, 7, 2, 7, 4, 9, 3, 1, 9, 7, 4, 1, 1, 1, 4, 0,
       9, 9, 1, 2, 3, 5, 0, 3, 4, 1, 5, 4, 9, 3, 5, 6, 4, 0, 9, 6, 7, 0,
       9, 9, 4, 7, 3, 5, 2, 0, 6, 7, 5, 3, 9, 7, 1, 3, 3, 9, 3, 3, 1, 4,
       1, 1, 1, 7, 1, 6, 4, 6, 9, 5, 8, 3, 5, 2, 9, 7, 9, 8, 2, 4, 1, 5,
       2, 2, 3, 9, 0, 2, 0, 2, 1, 0, 5, 0, 6, 4, 2, 1, 9, 0, 9, 0, 6, 9,
       4, 4, 9, 1, 5, 6, 1, 8, 7, 0, 8, 6, 2, 0, 1, 2, 3, 8, 4, 4, 3, 5,
       7, 3, 7, 2, 0, 2, 0, 9, 2, 8, 6, 3, 4, 0, 6, 6, 6, 7, 1, 6, 1, 7,
       6, 0, 6, 3, 7, 5, 6, 2, 2, 9, 4, 4, 7, 3, 3,

In [15]:
# 모델 평가해보기
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.92      0.96        37
           1       0.91      0.91      0.91        34
           2       0.80      0.82      0.81        34
           3       0.80      0.88      0.83        40
           4       0.75      0.79      0.77        34
           5       0.97      0.91      0.94        32
           6       0.94      0.92      0.93        37
           7       0.90      0.93      0.91        40
           8       0.82      0.70      0.75        33
           9       0.76      0.82      0.79        39

    accuracy                           0.86       360
   macro avg       0.87      0.86      0.86       360
weighted avg       0.86      0.86      0.86       360



In [20]:
# Random Forest 사용해 보기
from sklearn.ensemble import RandomForestClassifier
random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.97      0.99        37
           1       0.97      0.94      0.96        34
           2       0.92      1.00      0.96        34
           3       0.95      0.97      0.96        40
           4       0.94      0.94      0.94        34
           5       0.94      0.97      0.95        32
           6       1.00      1.00      1.00        37
           7       0.93      0.97      0.95        40
           8       0.97      0.88      0.92        33
           9       0.97      0.92      0.95        39

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.96      0.96      0.96       360



In [21]:
# SVM 사용해 보기
from sklearn import svm
svm_model =svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)
print(classification_report(y_test,y_pred))


              precision    recall  f1-score   support

           0       1.00      1.00      1.00        37
           1       0.97      1.00      0.99        34
           2       0.97      1.00      0.99        34
           3       1.00      0.97      0.99        40
           4       1.00      0.94      0.97        34
           5       0.97      1.00      0.98        32
           6       1.00      1.00      1.00        37
           7       1.00      1.00      1.00        40
           8       0.91      0.97      0.94        33
           9       1.00      0.95      0.97        39

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



In [22]:
# SGD Classifier 사용해 보기
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.97      0.99        37
           1       0.93      0.79      0.86        34
           2       0.94      1.00      0.97        34
           3       0.97      0.95      0.96        40
           4       0.97      0.91      0.94        34
           5       0.86      1.00      0.93        32
           6       1.00      1.00      1.00        37
           7       1.00      0.95      0.97        40
           8       0.76      0.94      0.84        33
           9       1.00      0.90      0.95        39

    accuracy                           0.94       360
   macro avg       0.94      0.94      0.94       360
weighted avg       0.95      0.94      0.94       360



In [24]:
# Logistic Regression 사용해 보기
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.97      1.00      0.99        37
           1       0.91      0.91      0.91        34
           2       0.92      1.00      0.96        34
           3       1.00      0.95      0.97        40
           4       0.94      0.91      0.93        34
           5       0.94      0.94      0.94        32
           6       0.97      0.97      0.97        37
           7       0.97      0.95      0.96        40
           8       0.89      0.94      0.91        33
           9       0.97      0.92      0.95        39

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


### 모델등 평가 비교 : 정확도 관점 
  - Decision Tree : 0.86
  - Random Forest : 0.96
  - SVM : 0.98
  - SGD Classifier : 0.94
  - Logistic Regression : 0.95
##### 정확도 관점에서는 SVM 의 정확성이 가장 높으나, precision, recall 등의 평가지표를 병행해서 볼 필요 있음
##### 데이터의 편중현상이 심할 경우, 정확도 만 가지고 평가하기 어려운 경우 발생하기 때문